libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,250 @@
+        } else if (b == NOOPLOOP_HEADER2) {
+            _msgbuf[0] = b;
+            _msg_len = 1;
+            _crc_expected = b;
+            _state = ParseState::H54_FUNCTION_MARK;
+        }
+        break;
+
+    case ParseState::H55_FUNCTION_MARK:
+        if (b == NOOPLOOP_FUNCTION_MARK_NODE_FRAME2) {
+            _msgbuf[1] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::LEN_L;
+        } else {
+            _state = ParseState::HEADER;
+        }
+        break;
+
+    case ParseState::H54_FUNCTION_MARK:
+        if (b == NOOPLOOP_FUNCTION_MARK_SETTING_FRAME0) {
+            _msgbuf[1] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::SF0_PAYLOAD;
+        } else {
+            _state = ParseState::HEADER;
+        }
+        break;
+
+    case ParseState::LEN_L:
+        _msgbuf[2] = b;
+        _msg_len++;
+        _crc_expected += b;
+        _state = ParseState::LEN_H;
+        break;
+
+    case ParseState::LEN_H:
+        // extract and sanity check frame length
+        _frame_len = (uint16_t)b << 8 | _msgbuf[2];
+        if (_frame_len > NOOPLOOP_NODE_FRAME2_FRAMELEN_MAX) {
+            _state = ParseState::HEADER;
+        } else {
+            _msgbuf[3] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::NF2_PAYLOAD;
+        }
+        break;
+
+    case ParseState::NF2_PAYLOAD:
+        // add byte to buffer if there is room
+        if (_msg_len < NOOPLOOP_MSG_BUF_MAX) {
+            _msgbuf[_msg_len] = b;
+        }
+        _msg_len++;
+        if (_msg_len >= _frame_len) {
+            _state = ParseState::HEADER;
+            // check crc
+            if (b == _crc_expected) {
+                return MsgType::NODE_FRAME2;
+            } else {
+                return MsgType::INVALID;
+            }
+        } else {
+            _crc_expected += b;
+        }
+        break;
+
+    case ParseState::SF0_PAYLOAD:
+        // add byte to buffer if there is room
+        if (_msg_len < NOOPLOOP_MSG_BUF_MAX) {
+            _msgbuf[_msg_len] = b;
+        }
+        _msg_len++;
+        if (_msg_len >= 128) {
+            _state = ParseState::HEADER;
+            // check crc
+            if (b == _crc_expected) {
+                return MsgType::SETTING_FRAME0;
+            } else {
+                return MsgType::INVALID;
+            }
+        } else {
+            _crc_expected += b;
+        }
+        break;
+    }
+
+    return MsgType::INVALID;
+}
+
+void AP_Beacon_Nooploop::parse_node_frame2()
+{
+    // a message has been received
+    _last_update_ms = AP_HAL::millis();
+
+    // estimated precision for x,y position in meters
+    const float precision_x = _msgbuf[NOOPLOOP_NODE_FRAME2_PRECISION_X] * 0.01;
+    const float precision_y = _msgbuf[NOOPLOOP_NODE_FRAME2_PRECISION_Y] * 0.01;
+    const float pos_err = sqrtf(sq(precision_x)+sq(precision_y));
```

Comment:
We've seen some issues before where the EKF's estimate goes very bad if the error value sent into the EKF is unrealistically low.  We should probably add a MAX(pos_err, some-value) to ensure it's never less than a reasonable value.  I'm not sure what that value should be but perhaps 0.1 (10cm).